### PR TITLE
rework input system

### PR DESCRIPTION
**Pull request by matrix-marker**
more clean, less coupled, & more extendable

### CODE REVIEW

File: src/main/java/core/input/AIInputDevice.java [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
 import com.badlogic.gdx.math.Vector2;
+import core.worldEntities.types.characters.Character;
 
-public class AIInputDevice implements CharacterInputDevice {
+public class AIInputDevice {
 
-    @Override
-    public CharacterInput getCharacterInput() {
-        return new CharacterInput(new Vector2(-1, 0), false);
+    private final InputProvider<Character.Input> characterInputProvider = () -> new Character.Input(new Vector2(-1, 0), false);
```
Comment (mdawess) [inline src/main/java/core/input/AIInputDevice.java]: Nice! Could you please comment for readability

File: src/main/java/core/input/InputController.java [hunk 2]
```diff
@@ -1,78 +1,19 @@
...
-import java.util.HashMap;
-import java.util.UUID;
-
 /**
- * Use-case class which manages InputDevices for GameCharacters.
- * Routes different kinds of inputs to their respective use-cases.
+ * Controller class for InputDevices.
+ * Currently, it does not control all that much since AI Input is very simple.
  */
 public class InputController {
```
Comment (mdawess) [inline src/main/java/core/input/InputController.java:13]: The controller class should take in inputs, package them and then pass them down to the use case layer. Is this just meant to be an intermediary class? If so, could it be an interface?

File: src/main/java/core/input/InputManager.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package core.input;
+
+import java.util.HashSet;
+
+public class InputManager {
```
Comment (mdawess) [inline src/main/java/core/input/InputManager.java:8]: Is this meant to be a use case? Just want to confirm level. Also, please comment

File: src/main/java/core/input/InputMapping.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package core.input;
+
+public class InputMapping<T extends InputSchema> {
+
+    private final InputProvider<T> provider;
+    private final InputHandler<T> handler;
+
+    public InputMapping(InputProvider<T> provider, InputHandler<T> handler) {
```
Comment (mdawess) [inline src/main/java/core/input/InputMapping.java]: This seems to be awkwardly between layers. The provider would be a controller and the handler is a use case. Why not just directly pass arguments from the provider to the handler without the middleman

Comment (mdawess) [reply src/main/java/core/input/InputMapping.java]: I noticed input provider is an interface - why are you instantiating?

File: src/main/java/core/input/KeyboardInputDevice.java [hunk 5]
```diff
@@ -3,17 +3,21 @@
 import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.Input;
 import com.badlogic.gdx.math.Vector2;
+import core.worldEntities.types.characters.Character;
 
-public class KeyboardInputDevice implements CharacterInputDevice, HudInputDevice {
+public class KeyboardInputDevice {
```
Comment (mdawess) [inline src/main/java/core/input/KeyboardInputDevice.java:10]: Please comment for readability

File: src/main/java/core/input/KeyboardInputDevice.java [hunk 6]
```diff
@@ -3,17 +3,21 @@
...
+    private final InputProvider<Character.Input> characterInputProvider = () -> new Character.Input(getDirection(), getUsing());
+
+    public InputProvider<Character.Input> characterInputProvider() {
+        return characterInputProvider;
     }
 
-    @Override
-    public HudInput getHudInput() {
-        return new HudInput(getInventory(), getPauseMenu());
+    private final InputProvider<core.presenters.HUD.HudInput> hudInputInputProvider = () ->
```
Comment (mdawess) [inline src/main/java/core/input/KeyboardInputDevice.java]: Why not import to make the code cleaner

File: src/main/java/core/presenters/levels/LevelGameplayController.java [hunk 7]
```diff
@@ -32,36 +32,34 @@
...
     private HudPresenter hudPresenter;
+    private InputManager inputManager;
     private CameraManager cameraManager;
-    private InputController inputController;
     private CharacterManager characterManager;
     private WorldEntityManager entityManager;
     private UUID playerId;
 
     @Override
     public void show() {
```
Comment (mdawess) [inline src/main/java/core/presenters/levels/LevelGameplayController.java:42]: please comment this more

**Review by mdawess — CHANGES_REQUESTED** [review]
Great changes, just a few comments!

**Review by roylin2000 — COMMENTED** [review]
As Michael mentioned, more JavaDocs would be nice to have. Otherwise, everything seems to work fine, so I'm fine with this :D

File: src/main/java/core/input/InputController.java (hunk 2 above)
Comment (matrix-marker) [reply src/main/java/core/input/InputController.java:13]: definitely worth explaining:

if in the future, AI inputs will depend on the state of the world, this controller class would take care of providing that input by storing the LevelManager, for example

File: src/main/java/core/input/AIInputDevice.java (hunk 1 above)
Comment (matrix-marker) [reply src/main/java/core/input/AIInputDevice.java]: this is a shorthand that I didn't even know existed until intellij prompted me to change what I had before

reworking this led me to do a Big Refactor which helped a Lot

File: src/main/java/core/presenters/levels/LevelGameplayController.java (hunk 7 above)
Comment (matrix-marker) [reply src/main/java/core/presenters/levels/LevelGameplayController.java:42]: not sure what's unclear; I think this class has the least changes in the whole PR

could you be more specific?

**Review by matrix-marker — COMMENTED** [review]
much improved over previous version imo

could still be tested using headless tests class

File: src/main/java/core/input/InputController.java (hunk 2 above)
Comment (mdawess) [reply src/main/java/core/input/InputController.java:13]: makes sense

**Review by mdawess — APPROVED** [review]

**Pull request merged**